plugins/basic/nexus-repository-docker/src/main/java/org/sonatype/nexus/repository/docker/internal/DockerHostedRecipe.groovy

```diff
@@ -0,0 +1,370 @@
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Provider
+import javax.inject.Singleton
+
+import org.sonatype.nexus.repository.Format
+import org.sonatype.nexus.repository.RecipeSupport
+import org.sonatype.nexus.repository.Repository
+import org.sonatype.nexus.repository.Type
+import org.sonatype.nexus.repository.docker.DockerConnectorFacet
+import org.sonatype.nexus.repository.docker.DockerFacet
+import org.sonatype.nexus.repository.http.HttpHandlers
+import org.sonatype.nexus.repository.partial.PartialFetchHandler
+import org.sonatype.nexus.repository.search.SearchFacet
+import org.sonatype.nexus.repository.security.SecurityHandler
+import org.sonatype.nexus.repository.storage.StorageFacetImpl
+import org.sonatype.nexus.repository.storage.UnitOfWorkHandler
+import org.sonatype.nexus.repository.types.HostedType
+import org.sonatype.nexus.repository.view.ConfigurableViewFacet
+import org.sonatype.nexus.repository.view.Context
+import org.sonatype.nexus.repository.view.Matcher
+import org.sonatype.nexus.repository.view.Route
+import org.sonatype.nexus.repository.view.Router
+import org.sonatype.nexus.repository.view.ViewFacet
+import org.sonatype.nexus.repository.view.handlers.ConditionalRequestHandler
+import org.sonatype.nexus.repository.view.handlers.TimingHandler
+import org.sonatype.nexus.repository.view.matchers.ActionMatcher
+import org.sonatype.nexus.repository.view.matchers.LiteralMatcher
+import org.sonatype.nexus.repository.view.matchers.PrefixMatcher
+import org.sonatype.nexus.repository.view.matchers.logic.LogicMatchers
+import org.sonatype.nexus.repository.view.matchers.token.TokenMatcher
+
+import org.eclipse.aether.util.version.GenericVersionScheme
+import org.eclipse.aether.version.InvalidVersionSpecificationException
+import org.eclipse.aether.version.VersionScheme
+
+import static org.sonatype.nexus.repository.http.HttpMethods.GET
+import static org.sonatype.nexus.repository.http.HttpMethods.POST
+import static org.sonatype.nexus.repository.http.HttpMethods.PUT
+
+/**
+ * Docker hosted repository recipe.
+ *
+ * @since 3.0
+ */
+@Named(DockerHostedRecipe.NAME)
+@Singleton
+class DockerHostedRecipe
+    extends RecipeSupport
+{
+  public static final String NAME = 'docker-hosted'
+
+  private static final VersionScheme SCHEME = new GenericVersionScheme()
+
+  private static final VERSION_0_10 = SCHEME.parseVersion('0.10')
+
+  @Inject
+  Provider<DockerSecurityFacet> securityFacet
+
+  @Inject
+  Provider<ConfigurableViewFacet> viewFacet
+
+  @Inject
+  Provider<DockerConnectorFacet> dockerConnectorFacet
+
+  @Inject
+  Provider<DockerFacet> dockerFacet
+
+  @Inject
+  Provider<StorageFacetImpl> storageFacet
+
+  @Inject
+  Provider<SearchFacet> searchFacet
+
+  @Inject
+  TimingHandler timingHandler
+
+  @Inject
+  SecurityHandler securityHandler
+
+  @Inject
+  PartialFetchHandler partialFetchHandler
+
+  @Inject
+  ConditionalRequestHandler conditionalRequestHandler
```

Comment:
This handler requires HTTP entity headers. So, either add `ContentHeadersHandler` too (to have it populate them for you), or add manually them somewhere.